runtime/tutorials/chat_app.md

```diff
@@ -163,73 +140,154 @@ function broadcast_usernames() {
+const socket = new WebSocket(
+  `ws://localhost:8080/start_web_socket?username=${myUsername}`,
+);
+
+// Handle WebSocket messages
+socket.onmessage = (event) => {
+  const data = JSON.parse(event.data);
+
+  switch (data.event) {
+    case "update-users":
+      updateUserList(data.usernames);
+      break;
+
+    case "send-message":
+      addMessage(data.username, data.message);
+      break;
   }
-  socket.username = username;
-  connectedClients.set(username, socket);
-  console.log(`New client connected: ${username}`);
-
-  // broadcast the active users list when a new user logs in
-  socket.onopen = () => {
-    broadcast_usernames();
-  };
-
-  // when a client disconnects, remove them from the connected clients list
-  // and broadcast the active users list
-  socket.onclose = () => {
-    console.log(`Client ${socket.username} disconnected`);
-    connectedClients.delete(socket.username);
-    broadcast_usernames();
-  };
-
-  // broadcast new message if someone sent one
-  socket.onmessage = (m) => {
-    const data = JSON.parse(m.data);
-    switch (data.event) {
-      case "send-message":
-        broadcast(
-          JSON.stringify({
-            event: "send-message",
-            username: socket.username,
-            message: data.message,
-          }),
-        );
-        break;
-    }
-  };
-});
+};
 
-app.use(router.routes());
-app.use(router.allowedMethods());
-app.use(async (context) => {
-  await context.send({
-    root: `${Deno.cwd()}/`,
-    index: "public/index.html",
+// Update user list in the DOM
+function updateUserList(usernames) {
+  const userList = document.getElementById("users");
+  userList.innerHTML = ""; // Clear existing list
+
+  for (const username of usernames) {
+    const listItem = document.createElement("li");
+    listItem.textContent = username;
+    userList.appendChild(listItem);
+  }
+}
+
+// Add a new message to the conversation
+function addMessage(username, message) {
+  const conversation = document.getElementById("conversation");
+  const messageDiv = document.createElement("div");
+  messageDiv.innerHTML = `<span>${username}</span> ${message}`;
+  conversation.prepend(messageDiv);
+}
+
+// Set up event listener for sending messages
+window.onload = () => {
+  const inputElement = document.getElementById("data");
+  const sendButton = document.getElementById("send");
+
+  // Listen for Enter key press
+  inputElement.addEventListener("keypress", (e) => {
+    if (e.key === "Enter") {
+      const message = inputElement.value;
+      inputElement.value = "";
+      socket.send(JSON.stringify({ event: "send-message", message }));
+    }
   });
-});
 
-console.log("Listening at http://localhost:" + port);
-await app.listen({ port });
+  // Listen for button click
+  sendButton.addEventListener("click", (e) => {
+    e.preventDefault();
+    const message = inputElement.value;
+    inputElement.value = "";
+    socket.send(JSON.stringify({ event: "send-message", message }));
+  });
```

Comment:
We should put the input element in a form and do normal form submission. That way we don't need to duplicate this logic and only need one `submit` listener.